impl/src/metadata.rs

```diff
@@ -63,34 +135,139 @@ impl MetadataFetcher for CargoMetadataFetcher {
     assert!(files.lock_path_opt.as_ref().map_or(true, |p| p.is_file()));
 
-    // Copy files into a temp directory
+    let tempdir = TempDir::new()?;
+
     // UNWRAP: Guarded by function assertion
-    let cargo_tempdir = {
-      let dir = TempDir::new()?;
-
-      let dir_path = dir.path();
-      let new_toml_path = dir_path.join(files.toml_path.file_name().unwrap());
-      fs::copy(files.toml_path.as_path(), new_toml_path)?;
-      if let Some(lock_path) = files.lock_path_opt.as_ref() {
-        let new_lock_path = dir_path.join(lock_path.file_name().unwrap());
-        fs::copy(lock_path.as_path(), new_lock_path)?;
-      }
-
-      dir
+    let cargo_dir = match self.use_tempdir {
+      false => files.toml_path.as_path().parent().unwrap(),
+      true => {
+        let dir_path = tempdir.path();
+        let new_toml_path = dir_path.join(files.toml_path.file_name().unwrap());
+        fs::copy(files.toml_path.as_path(), new_toml_path)?;
+        if let Some(lock_path) = files.lock_path_opt.as_ref() {
+          let new_lock_path = dir_path.join(lock_path.file_name().unwrap());
+          fs::copy(lock_path.as_path(), new_lock_path)?;
+        }
+
+        dir_path
+      },
     };
 
     MetadataCommand::new()
       .cargo_path(&self.cargo_bin_path)
-      .current_dir(cargo_tempdir.path())
+      .current_dir(cargo_dir)
       .exec()
       .map_err(|e| e.into())
   }
 }
 
+/**
+ * Ensures a lockfile is generated for a crate on disk
+ *
+ * If a lockfile for the current crate exists in the `raze` lockfiles output directory, that
+ * lockfile will instead be installed into the crate's directory. If no lockfile exists, one
+ * will be generated.
+ */
+fn cargo_generate_lockfile(
+  crate_dir: &Path,
+  lockfile_dir: &Path,
+  name: &str,
+  version: &str,
+) -> Result<PathBuf> {
+  let target_lockfile = lockfile_dir.join(format!("Cargo.{}-{}.lock", name, version));
+  let bundled_lockfile = crate_dir.join("Cargo.lock");
+
+  // If a lockfile exists from a previous run, use that instead
+  if target_lockfile.exists() {
+    fs::copy(&target_lockfile, &bundled_lockfile)?;
+    return Ok(bundled_lockfile);
+  }
+
+  // Generate a new lockfile if one wasn't included in the bundle
+  if !bundled_lockfile.exists() {
+    std::process::Command::new(SYSTEM_CARGO_BIN_PATH)
+      .arg("generate-lockfile")
+      .current_dir(&crate_dir)
+      .output()?;
+  }
+
+  Ok(bundled_lockfile)
+}
+
+pub fn fetch_crate_checksum(index_url: &str, name: &str, version: &str) -> Result<String> {
+  let index = crates_index::BareIndex::from_url(index_url)?;
+
+  let crate_index = index
+    .open_or_clone()?
+    .crate_(name)
+    .ok_or(anyhow!("Failed to find crate '{}' in index", name))?;
+
+  let crate_version = crate_index
+    .versions()
+    .iter()
+    .enumerate()
+    .find(|(_, ver)| ver.version() == version)
+    .ok_or(anyhow!(
+      "Failed to find version {} for crate {}",
+      version,
+      name
+    ))?
+    .1;
```

Comment:
Do this `.1` in a separate variable declaration.

Reply:
Done